Terminal.Gui/ConsoleDrivers/WindowsDriver.cs

```diff
@@ -806,7 +809,7 @@ public WindowsDriver ()
 		if (!RunningUnitTests) {
 			_parentProcessName = GetParentProcessName ();
 			_isWindowsTerminal = _parentProcessName == "WindowsTerminal";
-			if (!_isWindowsTerminal && _parentProcessName != "devenv") {
+			if (!_isWindowsTerminal && _parentProcessName != "wininit") {
```

Comment:
Maybe you can change to this:
```cs
			if (!_isWindowsTerminal && _parentProcessName != "wininit" && _parentProcessName != "devenv") {
				Force16Colors = true;
			}

```

Reply:
In line 1758 you can change to:
```cs
		if (!RunningUnitTests && _isWindowsTerminal) {
			// Disable alternative screen buffer.
			Console.Out.Write (EscSeqUtils.CSI_RestoreCursorAndActivateAltBufferWithBackscroll);
		}

```

Reply:
On the machine that was (for some reason!) returning "wininit" I'm now getting "devenv" again:

<img width="527" alt="image" src="https://github.com/gui-cs/Terminal.Gui/assets/585482/156061d2-647c-41e4-bcf5-8dc890a9db9b">

Reply:
On my main dev machine, I get "WindowsTerminal":

![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/8c8d6200-e017-46ad-8c02-8db3baec44b5)
![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/843f53c1-a1c7-4b23-8f8a-094307c20b1e)
![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/ac10ab35-0868-466d-a77d-96c843db97ae)
![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/bd7455f0-c29c-44ec-8948-50d8dc1b659b)

Reply:
It's pretty clear to me that relying on the process name is not going to work reliably.

Reply:
Related: https://github.com/microsoft/terminal/issues/13006

Reply:
> On the machine that was (for some reason!) returning "wininit" I'm now getting "devenv" again:
> 
> <img alt="image" width="527" src="https://user-images.githubusercontent.com/585482/276915125-156061d2-647c-41e4-bcf5-8dc890a9db9b.png"> <img alt="image" width="187" src="https://user-images.githubusercontent.com/585482/276915243-cebe820d-663b-4d5a-b578-fb32c4823b52.png"> <img alt="image" width="142" src="https://user-images.githubusercontent.com/585482/276915286-d0349147-2b48-4f8d-a32f-31b2a30d0eb9.png"> <img alt="image" width="249" src="https://user-images.githubusercontent.com/585482/276915379-8dd736c7-ac30-4faf-aa9a-8f1eccb81b92.png">

"devenv" is what also I get when debugging from VS2022.


> On my main dev machine, I get "WindowsTerminal":
> 
> ![image](https://user-images.githubusercontent.com/585482/276915669-8c8d6200-e017-46ad-8c02-8db3baec44b5.png) ![image](https://user-images.githubusercontent.com/585482/276915816-843f53c1-a1c7-4b23-8f8a-094307c20b1e.png) ![image](https://user-images.githubusercontent.com/585482/276915753-ac10ab35-0868-466d-a77d-96c843db97ae.png) ![image](https://user-images.githubusercontent.com/585482/276915725-bd7455f0-c29c-44ec-8948-50d8dc1b659b.png)

"WindowsTerminal" is what I get when running from `Windows Terminal` and attaching it to the VS2022 debugger.


![image](https://github.com/gui-cs/Terminal.Gui/assets/13117724/e7c0ea16-fd08-4b16-a656-1bb3a40083ee)
![image](https://github.com/gui-cs/Terminal.Gui/assets/13117724/4597a20c-c5ea-42df-ae16-00905fe19a8d)
![image](https://github.com/gui-cs/Terminal.Gui/assets/13117724/9dc533f2-c1c5-40b4-8a93-724cd927a388)

I'm not in the insiders program nor using preview versions, by option.

Reply:
> "WindowsTerminal" is what I get when running from Windows Terminal and attaching it to the VS2022 debugger.

I'm not attaching to a process. I'm hitting F5 in VS2022. 

Can we please agree that the right solution is to check for WS_SESSION and that looking at the parent process name is a hack that will not work long term? 

It is clear that WT_SESSION is also not reliable, but at least it's the method sanctioned by the WT team (and there's an open issue to fix it: https://github.com/microsoft/terminal/issues/13006)

In the meantime, instead of the hack to find the process name, a workaround for running in the VS2022 debugger is to change the launch profile thusly:

![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/ccfac2f9-ef46-470f-8146-fbf0053f9d9b)

I still don't understand why on one of my machines WT_SESSION is set correctly by default, but on another it is not.

Reply:
Also, I've found that `GetProcessName()` infinite loops:

- Start.Run "pwsh"
- `.\UICatalog\bin\Debug\net7.0\UICatalog.exe`
- Hangs
- Attach to debugger from VS 2022 and hit pause

It's stuck in the `while (queryObj != null)` loop.

Reply:
@tig the better way to check what terminal default is set is read Windows registry, see https://support.microsoft.com/en-us/windows/command-prompt-and-windows-powershell-for-windows-11-6453ce98-da91-476f-8651-5c14d5777c20.

In this case we don't need to get the process name and the method can be removed.

Reply:
Ok I saw your post, in this case we need to use the method to get the process name. I haven't any hang on running the method.

Reply:
> Ok I saw your post, in this case we need to use the method to get the process name. I haven't any hang on running the method.

Let's make sure you and I are talking about solving the same problem. Here's the problem I think we're trying to solve:

"When I run a Terminal.Gui app in the debugger in VS 2022 I want the WindowsDriver to support True Color. However, some versions of VS don't set WT_SESSION which is the correct way for Terminal.Gui to detect WT. We need a work around."

If this is indeed the problem, the correct fix is the code that is currently in this PR:

```cs
public WindowsDriver ()
{
	if (Environment.OSVersion.Platform == PlatformID.Win32NT) {
		WinConsole = new WindowsConsole ();
		// otherwise we're probably running in unit tests
		Clipboard = new WindowsClipboard ();
	} else {
		Clipboard = new FakeDriver.FakeClipboard ();
	}

	if (!RunningUnitTests) {
		_isWindowsTerminal = Environment.GetEnvironmentVariable ("WT_SESSION") != null;
		if (!_isWindowsTerminal) {
			//_parentProcessName = GetParentProcessName ();
			//_isWindowsTerminal = _parentProcessName == "WindowsTerminal";
			//if (!_isWindowsTerminal && _parentProcessName != "devenv") {
			Force16Colors = true;
			//}
		}
	}
}
```

PLUS, advice to change `launchSettings.json` to set the WT_SESSION env var:

```json
{
  "profiles": {
    "UICatalog": {
      "commandName": "Project",
      "environmentVariables": {
        "WT_SESSION": "yes"
      }
    },
...
```

This solves the problem.

Is this the problem you are talking about?

Reply:
> Is this the problem you are talking about?

Not really, your workaround is suffice when you are debugging from VS2022 by pressing F5 and that work well because VS2022 will always support true color.
I think relying on "WT_SESSION" isn't ensured because if you have others terminals open besides "Windows Terminal" the environment variable will always be available and then you don't have certain if is your application or not that is running on "Windows Terminal". In this case the `GetProcessName` method is more accurate. It's my opinions, for now to debug, your solution is fine.

Reply:
> It's pretty clear to me that relying on the process name is not going to work reliably.

For debug mode you are right, but for release mode is more reliable. Put inside an #if definition.

Reply:
In what scenario for the release version is there a problem that requires this check?

Reply:
> > Is this the problem you are talking about?
> 
> Not really, your workaround is suffice when you are debugging from VS2022 by pressing F5 and that work well because VS2022 will always support true color. I think relying on "WT_SESSION" isn't ensured because if you have others terminals open besides "Windows Terminal" the environment variable will always be available and then you don't have certain if is your application or not that is running on "Windows Terminal". In this case the `GetProcessName` method is more accurate. It's my opinion, for now to debug, your solution is fine.

I don't understand your concern. If the user runs a Terminal.Gui app on Windows in a terminal besides Windows Terminal, WT_SESSION will not be set. Environment variables can be either per-process or system wide. In this case we're talking about the WT process setting WT_SESSION, so it will only be scoped to that process.

Reply:
> In what scenario for the release version is there a problem that requires this check?

Easy. Run under cmd or conhost directly. If true is enabled you'll see a black and white screen with a bunch of Ansi escape sequences.

Reply:
Sorry @tig, it seems that you already fix it. Now `_isWindowsTerminal` is false by default and that fix it.